reed_solomon.go

```diff
@@ -0,0 +1,154 @@
+package chunk
+
+import (
+	"bytes"
+	"io"
+	"io/ioutil"
+	"sync"
+
+	"github.com/ipfs/go-ipfs-files"
+	rs "github.com/klauspost/reedsolomon"
+)
+
+const (
+	DefaultReedSolomonDataShards   = 10
+	DefaultReedSolomonParityShards = 20
+	DefaultReedSolomonShardSize    = DefaultBlockSize
+)
+
+// reedSolomonSplitter implements the MultiSplitter interface and splits into multiple
+// Splitters based on data + parity shards. Each Splitter corresponds to one
+// Reed-Solomon shard and splits using the default SizeSplitter.
+// Reed-Solomon shard Splitters are safe for concurrent reading.
+// The default Splitter interface for ReedSolomonSplitter is a serialized
+// read of all shard chunks.
+type reedSolomonSplitter struct {
+	sync.Mutex
+
+	r         io.Reader
+	spls      []Splitter
+	splIndex  int
```

Comment:
splIndex is a shared variable and is not read only. I think we need to synchronize access operations like NextBytes() .

Reply:
Each individual splitter's NextBytes is still thread safe because they don't share variables.

Reply:
As a reference, https://github.com/TRON-US/go-btfs-chunker/blob/master/splitting.go#L90 is not thread safe because it must be called serialized

Reply:
Will add a comment on next commit